### ISSUE TITLE
Breakpoints don't work when the workspace folder is a symlink

### ISSUE THREAD

**Issue by DanTup** [Dart-Code/Dart-Code#3896]
Raised by @nmfisher at https://github.com/Dart-Code/Dart-Code/issues/3808#issuecomment-1084436774

---

I've been tearing my hair out today trying to understand why everything worked fine in a Docker image I put together, but still not in the original Linux environment that was very similar.

That's until I realized that in my original environment, VS Code Remote SSH was actually opening the Flutter project folder at a symbolic link, not the actual filesystem location. I created a sample project on Linux to check, and it seems that indeed, breakpoints do not work if folders are opened via symbolic link.

So forget about port forwarding/Remote SSH, I'm pretty sure the underlying problem can be replicated as follows (all on Linux):

1. Run the Android emulator.
2. Run the following on the command line:
```
cd ~
flutter create flutter_test_project
ln -s flutter_test_project $(pwd)/flutter_test_project_link
```
3. Run VS Code and open `~/flutter_test_project` folder
4. Set a breakpoint at `_incrementCounter` line 54 in main.dart, hit F5 run
5. In emulator, once the app is running, click the + floating action button, VS Code should pause/highlight on the breakpoint as expected
7. In VS Code, close the `~/flutter_test_project` folder and open the `~/flutter_test_project_link` folder
8. Set the same breakpoint as (4), hit F5 run
9. In the emulator, click the + button and see that VS Code does not pause on the breakpoint. 

I confirm that opening the remote SSH folder at the actual filesystem location, not the symbolic link, resolves my issue and everything is working as intended.

**Comment by DanTup** [comment]
I can repro this in an Ubuntu container. It seems that Flutter or Dart is resolving the symlinks to their underlying paths, which means the paths that VS Code sends (for example for breakpoints) don't match the scripts that are loaded, and the breakpoints not hit.

A simpler repro is to just run `flutter test --machine` in a symlinked folder - the test paths you'll see printed are for the real location, not the symlink:

```
vscode ➜ ~/flutter_test_project_link $ pwd && ~/flutter/bin/flutter test --machine
/home/vscode/flutter_test_project_link

{"suite":{"id":0,"platform":"vm","path":"/workspaces/flutter_test_project/test/widget_test.dart"},"type":"suite","time":0}
```

Eliminating Flutter, if I just make a simple Dart script and run that, I see the same:

```
[1:48:59 PM] [VmService] [Info] [Dart ] Spawning /home/vscode/flutter/bin/cache/dart-sdk/bin/dart with args ["--enable-vm-service=0","bin/dart_test.dart"]
[1:48:59 PM] [VmService] [Info] [Dart ] ..  in /home/vscode/dart_test_link
```

It was executed in the symlink folder, but the VM Service shows the scripts loaded from the real locations:

```
"uri":"file:///home/vscode/dart_test/bin/dart_test.dart"
```

@bkonyi do you know the reason this happens? This sounds familiar like we discussed it before, but I wasn't able to find any issues that seem related.

**Comment by DanTup** [comment]
@bkonyi do you know the reasons for this (or who might)? I'm not sure whether this is a bug or expected (and if it's expected, whether Dart-Code/DAP will need to work around it).

The issue is that if you create a symlink folder and then run a Dart script from it, the paths from the VM Service are all _resolved_. This means if the IDE/DAP hasn't already previously resolved all paths, they won't match up.

**Comment by bkonyi** [comment]
Oh, yeah, I'm guessing that the frontend resolves the symlinks to real paths and puts those in the generated kernel for the program, so the VM has no clue about the original symlink path which results in breakpoints not being set properly.

I don't know 100% that's what's actually happening, but maybe someone on the frontend team will know.

**Comment by johnniwinther** [comment]
The frontend works through a file-system abstraction that doesn't show the underlying system resource, so the frontend only knows about URIs which it normal get directly from the user code itself.

A possible culprit could be the package config system, which is responsible for converting `package:` URIs into `file:` URIs. @lrhn do you know how `package_config` handles symbolic links?

**Comment by lrhn** [comment]
The `package_config` package doesn't handle symbolic links. It's completely ignorant of them, and just works symbolically with paths and (mostly) URIs.

I don't think it does any attempt to get to the physical path at any time. (There is no occurrence of either of the words "symbolic" or "link" in the repository.)

So, what it does is to use the *entry point path* as base, look in `.dart_tool` for `package_config.json`, and if not finding it or `.packages`, remove the last part of the base part and try again.

That means it matters is what entry point path it is given, which depends on who's calling. If the front-end/VM resolves symbolic links before passing in the entry point path, then that's what it's going to work with.

When `package_config` finds a `package_config.json` file, it creates a URI from the path of the file, then resolves the package and `lib` paths of the configuration against that URI. That's all relative URI resolution *unless the `package_config.json` contains absolute paths!* 
So, check for that.

**Comment by DanTup** [comment]
I tried creating a sample app at `./real_app`, deleted `.dart_tool` and `.packages`. I then created a `./linked_app` symlink that points at it, then ran `dart pub get` from inside `./linked_app`.

The newly generated `.dart_tool/package_config.json` just has normal relative paths:

```
{
  "name": "the_app",
  "rootUri": "../",
  "packageUri": "lib/",
  "languageVersion": "2.16"
}
```

And the `.packages` file contains this:

```
the_app:lib/
```

So there's nothing _on disk_ that I can see that has resolved this link in any way. But if I add `print('Hello world: ${Directory.current}!');` and run `dart bin/the_app.dart` (from inside `linked_app`) it prints:

> Hello world: Directory: '/Users/danny/Desktop/dart_samples/dartcode_3869/real_app'!

And if I run with `--observe`, Observatory shows the same `real_app` as the root URI for the isolate:

![Screenshot 2022-05-02 at 14 18 20](https://user-images.githubusercontent.com/1078012/166240202-8c633da1-eee4-4f34-ab25-9f45857c52e3.png)

**Comment by bkonyi** [comment]
For Directory.current, we use getcwd
<https://pubs.opengroup.org/onlinepubs/007904975/functions/getcwd.html>
which states that "pathname copied to the array shall contain no components
that are symbolic links", so that behavior is expected.

One way we can check whether or not the library path is coming from the CFE
or somehow being resolved by the VM is to dump the kernel file and look at
the paths associated with it. I did some testing locally with the following
directory structure at ~:

testing/
  bin/
    main.dart
sym_testing/ -> testing/

And ran the following:

cd ~/sym_testing/
# cwd -> /Users/bkonyi/testing_sym
dart compile kernel bin/main.dart

While we have a tool to dump the kernel structure in a more readable form
to disk, opening the generated main.dill is easy enough to interpret for a
program this small. All of the paths for the library I see in the dill file
are of the form file:///Users/bkonyi/testing/bin/main.dart, which tells me
the VM isn't doing anything weird with the paths after loading the dill
file. Something is happening when we generate the kernel, which I'm
guessing is related to the system level file system APIs we use to back the
Dart implementations.

At this point, I'm not sure this is a bug, and we should specifically
specify that breakpoints should only be set using absolute paths. Although,
I guess we could always try and resolve the paths in the VM if breakpoint
creation fails and try again, but this wouldn't work on Flutter since the
kernel is generated on a different device.

**Comment by DanTup** [comment]
> At this point, I'm not sure this is a bug, and we should specifically
> specify that breakpoints should only be set using absolute paths

I'm also not sure it's a bug, but I wanted to find out for sure because I think fixing it in the IDE/DAP might be complicated (or even not possible without edge cases)..

For DAP we don't have any control over the editor, so incoming requests to set breakpoints (etc.) may contain the symlinked paths. While we can easily map/resolve those, it's less clear what we do going the other way.. if we get a resolved path from the VM (eg. when we pause on an exception), how do we know if we should map it back? We don't know about the symlink unless we previously resolved a path in the other direction, and even then it's not guaranteed to be a one-to-one mapping.

If we can't support this reliably and have to consider it a known limitation, it would be good to at least know the source (and reasons) for it.

**Comment by johnniwinther** [comment]
Another possible source is the package_config.json file itself, which is often generated by pub. @jonasfj do you know if pub resolves symlinks when generating the package_config.json file?

**Comment by DanTup** [comment]
In my testing above (https://github.com/Dart-Code/Dart-Code/issues/3896#issuecomment-1114868074), the `package_config.json` had relative paths that did not appear to have resolved symlinks.

**Comment by jonasfj** [comment]
Yeah, `path`-dependencies are often relative paths, where as dependencies from `PUB_CACHE` are generally absolute. I'm not sure we resolve symlinks, but it's entirely possible.

**Comment by DanTup** [comment]
@bkonyi @johnniwinther I'm not sure how to progress this (or even confirm whether it's a bug or intended). The steps above at https://github.com/Dart-Code/Dart-Code/issues/3896#issuecomment-1114868074 reproduce this fairly easily (showing that the `Directory.current` is the resolved folder even though the app is run from the symlink, and `package_config.json` does not reference the real folder.

If it's deemed that this is WAI then we could consider trying to detect this scenario and warn the user, but since this might prevent some setups from working correctly (for ex. using VS Code w/ Docker remoting) it'd be good to confirm this isn't just a bug/oversight somewhere.

**Comment by bkonyi** [comment]
AFAICT the VM is WAI given what information it has access to. I think whatever fix we come up with here will need to be somewhere upstream (e.g., the CFE or `package_config`).

**Comment by bkonyi** [comment]
Yeah, that might be a good idea. I'd cc Slava and Siva as well since they also have a good understanding of the various components involved here.

**Comment by DanTup** [comment]
This came up at https://github.com/Dart-Code/Dart-Code/issues/4172 too. It also affects setting the root directory for DevTools.

**Comment by johnniwinther** [comment]
@DanTup Given that https://github.com/dart-lang/sdk/issues/49309 has been closed, do we know if the fix hasn't made it to the right channel yet, if the fix was insufficient, or if we are seeing a new context in which this error occurs?

**Comment by DanTup** [comment]
Looks like passing the full path as the target is not enough. I tried this:

```
cd ~/Dev/TestStuff/
flutter create flutter_test_project
ln -s flutter_test_project $(pwd)/flutter_test_project_link
cd flutter_test_project_link
flutter pub get
cat .dart_tool/package_config.json
flutter run --target $(pwd)/lib/main.dart -d flutter-tester -v | tee output.txt
```

In `output.txt` I see a mix of `flutter_test_project` and `flutter_test_project_link` (I wanted to only see `flutter_test_project_link`). An example of a "bad" path is:

```
[  +29 ms] /Users/danny/Dev/Google/flutter/bin/cache/artifacts/engine/darwin-x64/flutter_tester --run-forever --non-interactive --enable-dart-profiling --packages=/Users/danny/Dev/TestStuff/flutter_test_project/.dart_tool/package_config.json --flutter-assets-dir=/var/folders/b1/ftm2whhj4mb3_c4nfjhq23z00000gn/T/flutter_tools.5BAQaC/flutter_tester.d0Zakp /var/folders/b1/ftm2whhj4mb3_c4nfjhq23z00000gn/T/flutter_tools.5BAQaC/flutter_tester.d0Zakp/flutter-tester-app.dill.track.dill
```

When running Dart-Code's integration tests, many fail because paths we expect to be `package:foo` are not (they're file paths). I believe this is because they're not being mapped correctly because reading the `package_config.json` from the original (real) path means the relative paths inside it are resolved using the original (real) path and don't match the new (linked) paths that the entry point (`--target`) was loaded from.

I'll need to do some more digging to understand where this is coming from. I suspect this is Flutter-specific and that with these changes to use absolute paths that Dart stuff _will_ work correctly - however I don't want to land that change without knowing it wouldn't make anything in Flutter today _worse_ given the above.

**Comment by DanTup** [comment]
In `flutter_tools` there are a number of places that use `Directory.current` (or equiv wrappers) such as here:

https://github.com/flutter/flutter/blob/bc70ae0f57b4a22722523ec28945344fb3dd4468/packages/flutter_tools/lib/src/project.dart#L100

Hard-coding the symlink path there fixes many of the places that were using the resolved path, although there are still some remaining. Some of these could probably be updated to compute the project root from the passed `--target` and not using `currentDirectory`, but I don't know if that will solve all issues (or whether it might introduce others).

While investigating this I found that my understanding that VS Code was automatically always resolving symlinks on macOS was not correct. It seems like macOS is doing that when using the file dialog, or using "." in the terminal. If I explicitly pass a full absolute file path at the terminal, it opens the symlinked folder.

```
code . // resolves the symlink
code /Users/danny/Dev/TestStuff/flutter_test_project_link // retains the symlink
```

This means it's easier to reproduce this from the terminal on macOS by just passing a full path.

That said, I don't know how many people are affected/how common this setup is. I feel like completely fixing it is not going to be trivial (even if all places in Flutter are fixed, it's very possible some of the external tools it calls to compile for each platform may also end up resolving somewhere).

I wonder if it would be acceptable to show a warning when a workspace is opened that doesn't match the resolved path (like we do [if the casing is mismatched](https://github.com/Dart-Code/Dart-Code/issues/2361)) and encourage users to open the source folder. I don't know of any cases where opening the source folder would not be possible/not work.

**Comment by plasmasolutions** [comment]
Hi Dan,

as I was searching for countless hours now and finally stumbled across your issue, I just want to raise my hand and say: "I'm affected as well". I would have never been able to pin the issue down to the symlink, so thanks very much!  I'm now able to at least debug on chrome again (moved the project), but I still have to see why I'm not able debug on Linux as a target.

Have a great day, 
Thomas

**Comment by FMorschel** [comment]
I've run into this too. You can either clone https://github.com/fmorschel/repro or do the changes I did to `dart` create example:

![Image](https://github.com/user-attachments/assets/19242ebc-6506-460f-b5dc-9578fbd6435d)

Here are all the log files:

[repro_wsl_ubuntu.zip](https://github.com/user-attachments/files/20983082/repro_wsl_ubuntu.zip)

This only happened with a symlink inside WSL for me. I've never had any symlink projects on Linux, but I did on Windows, no issue.

The actual folder is under `/mnt/d/dev/dart/repro` and the symlink is at `~/dev/dev/dart/repro` (yes, two `dev`, I'm moving some stuff around) pointing to that. I'm opening the project/vscode at `~/dev/dev/dart/repro`. 

The repro actually has a difference from the above image in the test folder to show that the relative folder works fine. The problem is with `package` URIs.
